File: CommonCode/GLRenderer.m
//
//  GLRenderer.m
//  SpectrogeddonOSX
//
//  Created by Tom York on 25/04/2014.
//  Copyright (c) 2014 Spectrogeddon. All rights reserved.
//

#import "GLRenderer.h"
#import "RendererDefs.h"
#import "LinearScrollingRenderer.h"
#import "ColumnRenderer.h"
#import "RenderTexture.h"
#import "TimeSequence.h"
#import "DisplaySettings.h"
#import "ColorMap.h"

@interface GLRenderer ()
@property (nonatomic) NSTimeInterval lastDuration;
@property (nonatomic,strong) ColumnRenderer* channel1Renderer;
@property (nonatomic,strong) ColumnRenderer* channel2Renderer;
@property (nonatomic,strong) LinearScrollingRenderer* scrollingRenderer;
@property (nonatomic,strong) RenderTexture* renderTexture;

@property (nonatomic) NSTimeInterval frameOriginTime;
@property (nonatomic) NSTimeInterval lastRenderedSampleTime;
@property (nonatomic,strong) DisplaySettings* displaySettings;
@end

@implementation GLRenderer

- (instancetype)init
{
    if((self = [super init]))
    {
        _scrollingRenderer = [[LinearScrollingRenderer alloc] init];
        _renderTexture = [[RenderTexture alloc] init];
        _channel1Renderer = [[ColumnRenderer alloc] init];
        _channel2Renderer = [[ColumnRenderer alloc] init];
    }
    return self;
}

- (void)setRenderSize:(RenderSize)renderSize
{
    if(!RenderSizeEqualToSize(renderSize, _renderSize))
    {
        _renderSize = renderSize;
        self.frameOriginTime = 0;
        self.renderTexture.renderSize = [self transformedRenderSize];
    }
}

- (void)render
{
    if(RenderSizeIsEmpty(self.renderSize) || !self.displaySettings)
    {
        return;
    }
    
    const NSTimeInterval nowTime = CACurrentMediaTime();
    if(!self.frameOriginTime)
    {
        self.frameOriginTime = nowTime;
    }
    
    float position = [self widthFromTimeInterval:nowTime - self.frameOriginTime];
    if(position > 1.0f)
    {
        self.frameOriginTime = nowTime;
        position = 0.0f;
    }
    self.scrollingRenderer.scrollingPosition = position;
    
    glViewport(0, 0, self.renderSize.width, self.renderSize.height);
    glClearColor(0.0f, 0.0f, 1.0f, 1.0f);
    glClear(GL_COLOR_BUFFER_BIT);
    [self.renderTexture renderTextureWithCommands:^{
        [self.scrollingRenderer render];
    }];
    
    GL_DEBUG_GENERAL;
}

- (void)addMeasurementsForDisplay:(NSArray*)spectrums
{
    const BOOL showStereo = spectrums.count > 1;
    if(showStereo)
    {
        self.channel1Renderer.positioning = [self positionForChannelAtIndex:0 totalChannels:2];
        self.channel2Renderer.positioning = [self positionForChannelAtIndex:1 totalChannels:2];
        [self updateChannelRenderer:self.channel1Renderer withSequence:[spectrums firstObject]];
        [self updateChannelRenderer:self.channel2Renderer withSequence:[spectrums lastObject]];
    }
    else
    {
        self.channel1Renderer.positioning = [self positionForChannelAtIndex:0 totalChannels:1];
        [self updateChannelRenderer:self.channel1Renderer withSequence:[spectrums firstObject]];
    }
    id __weak weakSelf = self;
    [self.renderTexture drawWithCommands:^{
        
        GLRenderer* strongSelf = weakSelf;
        if(strongSelf)
        {
            [strongSelf.channel1Renderer render];
            if(showStereo)
            {
                [strongSelf.channel2Renderer render];
            }
            self.lastRenderedSampleTime = [[spectrums firstObject] timeStamp];
        }
    }];
}

- (void)useDisplaySettings:(DisplaySettings *)displaySettings
{
    self.displaySettings = displaySettings;
    self.channel1Renderer.colorMapImage = [displaySettings.colorMap imageRef];
    self.channel2Renderer.colorMapImage = [displaySettings.colorMap imageRef];
    self.channel1Renderer.useLogFrequencyScale = displaySettings.useLogFrequencyScale;
    self.channel2Renderer.useLogFrequencyScale = displaySettings.useLogFrequencyScale;
    self.renderTexture.renderSize = [self transformedRenderSize];
    self.scrollingRenderer.scrollVertically = self.displaySettings.scrollVertically;
}

- (RenderSize)transformedRenderSize
{
    return self.displaySettings.scrollVertically ? (RenderSize) { self.renderSize.height, self.renderSize.width } : self.renderSize;
}

- (GLKMatrix4)positionForChannelAtIndex:(NSUInteger)channelIndex totalChannels:(NSUInteger)totalChannels
{
    if(totalChannels == 0)
    {
        return GLKMatrix4Identity;
    }
    const float channelHeight = 2.0f/(float)(totalChannels);
    const BOOL flipChannel = (channelIndex & 1);    // Flip odd numbered channels.
    const float center = 1.0f - channelHeight*(float)(channelIndex + 1 - (flipChannel ? 1 : 0));
    GLKMatrix4 positioning = GLKMatrix4MakeTranslation(0.0f, center, 0.0f);
    positioning = GLKMatrix4Scale(positioning, 1.0f, channelHeight*(flipChannel ? -1.0f : 1.0f), 1.0f);
    return positioning;
}

- (void)updateChannelRenderer:(ColumnRenderer*)renderer withSequence:(TimeSequence*)timeSequence
{
    float baseOffset = [self widthFromTimeInterval:(timeSequence.timeStamp - self.frameOriginTime)];
    if(baseOffset > 1.0f)
    {
        baseOffset = 0.0f;
    }
    self.lastDuration = timeSequence.duration;
    const float width = [self widthFromTimeInterval:timeSequence.duration + timeSequence.timeStamp - self.lastRenderedSampleTime];
    [renderer updateVerticesForTimeSequence:timeSequence offset:(2.0f * baseOffset - 1.0f) width:width];
}

- (float)widthFromTimeInterval:(NSTimeInterval)timeInterval
{
    const float screenFractionPerSecond = self.displaySettings.scrollingSpeed/(self.lastDuration * (float)([self transformedRenderSize].width));
    return screenFractionPerSecond * timeInterval;
}

@end

File: CommonCode/GLRenderer.swift
//
//  GLRenderer.swift
//  SpectrogeddonOSX
//
//  Created by Tom York on 18/05/2019.
//  Copyright © 2019 Spectrogeddon. All rights reserved.
//

import Foundation
import GLKit

class GLRenderer {
    
    var namesForScrollingDirections: [String] {
        return scrollingRenderer.namesForScrollingDirections
    }

    var renderSize: RenderSize {
        didSet {
            if renderSize != oldValue {
                resetTimeline()
                renderTexture.renderSize = scrollingRenderer.bestRenderSize(for: renderSize)
            }

        }
    }
    
    private var channel1Renderer: ColumnRenderer
    private var channel2Renderer: ColumnRenderer
    private var scrollingRenderer: ScrollingRenderer
    private var renderTexture: RenderTexture
    private var displaySettings: DisplaySettings?

    private var lastDuration: TimeInterval
    private var lastMeasurementTime: TimeInterval
    private var lastMeasurementPosition: Float
    
    private var scrollingPositionNow: Float {
        guard lastMeasurementTime > 0 else {
            return 0
        }

        return normalizedFraction(
            TimeInterval(lastMeasurementPosition) + TimeInterval(widthFromTimeInterval(CACurrentMediaTime() - lastMeasurementTime))
        )
    }

    
    init() {
        self.renderSize = .init(width: 0, height: 0)
        self.scrollingRenderer = LinearScrollingRenderer()
        self.renderTexture = RenderTexture()
        self.channel1Renderer = ColumnRenderer()
        self.channel2Renderer = ColumnRenderer()
        lastDuration = 0
        lastMeasurementTime = 0
        lastMeasurementPosition = 0
    }

    func addMeasurements(_ channels: [TimeSequence]) {
        
        guard !channels.isEmpty else {
            return
        }
        
        // TODO: model this better
        let showStereo = channels.count > 1
        let currentMeasurementTime = CACurrentMediaTime()
        let measuredDuration = max(channels[0].duration, 0)
        let elapsedMeasurementTime = lastMeasurementTime > 0 ? currentMeasurementTime - lastMeasurementTime : measuredDuration
        let previousMeasurementPosition = lastMeasurementPosition
        lastDuration = measuredDuration
        let measuredWidth = widthFromTimeInterval(elapsedMeasurementTime)
        let renderSegments = splitRenderSegment(offset: previousMeasurementPosition, width: measuredWidth)

        if showStereo {
            channel1Renderer.positioning = positionForChannelAtIndex(channelIndex: 0, totalChannels: 2)
            channel2Renderer.positioning = positionForChannelAtIndex(channelIndex: 1, totalChannels: 2)
        } else {
            channel1Renderer.positioning = positionForChannelAtIndex(channelIndex: 0, totalChannels: 1)
        }

        renderTexture.draw { [weak self] in
            
            guard let strongSelf = self else {
                return
            }
            strongSelf.renderChannel(renderer: strongSelf.channel1Renderer, withSequence: channels[0], segments: renderSegments)
            if showStereo {
                strongSelf.renderChannel(renderer: strongSelf.channel2Renderer, withSequence: channels[1], segments: renderSegments)
            }
            strongSelf.lastMeasurementPosition = strongSelf.normalizedFraction(TimeInterval(previousMeasurementPosition + measuredWidth))
            strongSelf.lastMeasurementTime = currentMeasurementTime
        }
    }
    
    func render() {
        
        guard renderSize != .empty, displaySettings != nil else {
            return
        }
        
        scrollingRenderer.scrollingPosition = scrollingPositionNow
        glViewport(0, 0, renderSize.width, renderSize.height)
        glClearColor(0, 0, 1, 1)
        glClear(GLbitfield(GL_COLOR_BUFFER_BIT))
        renderTexture.renderTexture {
            self.scrollingRenderer.render()
        }
    }
    
    func use(_ displaySettings: DisplaySettings) {
        
        self.displaySettings = displaySettings
        channel1Renderer.colorMapImage = displaySettings.colorMap!.image
        channel2Renderer.colorMapImage = displaySettings.colorMap!.image
        channel1Renderer.useLogFrequencyScale = displaySettings.useLogFrequencyScale
        channel2Renderer.useLogFrequencyScale = displaySettings.useLogFrequencyScale
        scrollingRenderer.activeScrollingDirectionIndex = displaySettings.scrollingDirectionIndex
        let bestRenderSize = scrollingRenderer.bestRenderSize(for: renderSize)
        if renderTexture.renderSize != bestRenderSize {
            resetTimeline()
        }
        renderTexture.renderSize = bestRenderSize
    }
    
    private func positionForChannelAtIndex(channelIndex: UInt, totalChannels: UInt) -> GLKMatrix4 {
        
        guard totalChannels > 0 else {
            return GLKMatrix4Identity
        }
        
        let channelHeight: Float = 2 / Float(totalChannels)
        let flipChannel: Bool = (channelIndex & 1) == 1 // Flip odd numbered channels.
        let center: Float = 1 - channelHeight * Float(channelIndex + 1 - (flipChannel ? 1 : 0))
        let positioning: GLKMatrix4 = GLKMatrix4MakeTranslation(0, center, 0)
        return GLKMatrix4Scale(positioning, 1, channelHeight*(flipChannel ? -1 : 1), 1)
    }
    
    private func renderChannel(renderer: ColumnRenderer, withSequence timeSequence: TimeSequence, segments: [(offset: Float, width: Float)]) {

        for segment in segments {
            renderer.updateVertices(timeSequence: timeSequence, offset: (2 * segment.offset - 1), width: 2 * segment.width)
            renderer.render()
        }
    }
    
    private func widthFromTimeInterval(_ timeInterval: TimeInterval) -> Float {

        guard let textureCycleDuration = textureCycleDuration else {
            return 0
        }
        return Float(timeInterval / textureCycleDuration)
    }

    private var textureCycleDuration: TimeInterval? {

        guard
            let displaySettings = displaySettings,
            displaySettings.scrollingSpeed > 0,
            lastDuration > 0,
            renderTexture.renderSize.width > 0 else {
            return nil
        }

        return lastDuration * TimeInterval(renderTexture.renderSize.width) / TimeInterval(displaySettings.scrollingSpeed)
    }

    private func splitRenderSegment(offset: Float, width: Float) -> [(offset: Float, width: Float)] {

        guard width > 0 else {
            return []
        }

        if width >= 1 {
            return [(offset: 0, width: 1)]
        }

        let firstSegmentWidth = min(width, 1 - offset)
        var segments = [(offset: offset, width: firstSegmentWidth)]
        let remainingWidth = width - firstSegmentWidth
        if remainingWidth > 0 {
            segments.append((offset: 0, width: remainingWidth))
        }
        return segments
    }

    private func resetTimeline() {

        lastMeasurementTime = 0
        lastMeasurementPosition = 0
    }

    private func normalizedFraction(_ value: TimeInterval) -> Float {

        guard value.isFinite else {
            return 0
        }
        return Float(value - floor(value))
    }
}
